### ISSUE TITLE
(Oscar-3.0.93) "Index 47 out of bounds" selecting an existing method in the method browser

### ISSUE THREAD

**Issue by dalehenrich** [ericwinger/Jade#738]
```
***** Starting Dolphin stack dump at 11:07:23 AM 3/4/2020 *****

************************** Dolphin Virtual Machine Dump Report ***************************

11:07:23 AM, 3/4/2020: Index 47 is out of bounds

*----> VM Context <----*
Process: {1ECE0210:size 583 words, suspended frame 1ECE0CD5, priority 5, callbacks 2
last failure 1:nil, FP control 80017, thread 00000000}
Active Method: SessionManager>>logError:
IP: 1A8AF4B9 (9)
SP: 1ECE0C30
BP: 1ECE0C04 (622)
ActiveFrame: {1ECE0C08: cf 1ECE0BE9, sp 1ECE0C20, bp 1ECE0C04, ip 5, JadeSessionManager(SessionManager)>>logError:}
	receiver: a JadeSessionManager
	arg[0]: a BoundsError


New Method: VMLibrary>>dump:path:stackDepth:walkbackDepth:
Message Selector: dump:path:stackDepth:walkbackDepth:

*----> Stack Back Trace <----*
{1ECE0C08: cf 1ECE0BE9, sp 1ECE0C20, bp 1ECE0C04, ip 5, JadeSessionManager(SessionManager)>>logError:}
	receiver: a JadeSessionManager
	arg[0]: a BoundsError

{1ECE0BE8: cf 1ECE0BC5, sp 1ECE0BFC, bp 1ECE0BE0, ip 39, JadeSessionManager>>logError:}
	receiver: a JadeSessionManager
	arg[0]: a BoundsError
	stack temp[0]: '11:07:23 AM 3/4/2020'
	env temp[0]: a FileStream

{1ECE0BC4: cf 1ECE0BA5, sp 1ECE0BD8, bp 1ECE0BC0, ip 3, JadeSessionManager(SessionManager)>>unhandledException:}
	receiver: a JadeSessionManager
	arg[0]: a BoundsError

{1ECE0BA4: cf 1ECE0B85, sp 1ECE0BB8, bp 1ECE0BA0, ip 3, JadeSessionManager(SessionManager)>>onUnhandledError:}
	receiver: a JadeSessionManager
	arg[0]: a BoundsError

{1ECE0B84: cf 1ECE0B69, sp 1ECE0B98, bp 1ECE0B84, ip 5, BoundsError(Error)>>defaultAction}
	receiver: a BoundsError

{1ECE0B68: cf 1ECE0B35, sp 1ECE0B7C, bp 1ECE0B50, ip 55, BoundsError(Exception)>>_propagateFrom:}
	receiver: a BoundsError
	arg[0]: a ExceptionHandler
	stack temp[0]: nil
	stack temp[1]: a ExceptionHandler
	stack temp[2]: nil
	stack temp[3]: a Process('Main' base 1ECE0000 [ACTIVE] in SessionManager>>logError: sp=00000000 ip=4 list=02770010)
	stack temp[4]: nil

{1ECE0B34: cf 1ECE0B15, sp 1ECE0B48, bp 1ECE0B30, ip 6, BoundsError(Exception)>>_propagate}
	receiver: a BoundsError
	stack temp[0]: nil

{1ECE0B14: cf 1ECE0AF9, sp 1ECE0B28, bp 1ECE0B14, ip 13, BoundsError(Exception)>>signal}
	receiver: a BoundsError

{1ECE0AF8: cf 1ECE0AD5, sp 1ECE0B0C, bp 1ECE0AF0, ip 8, BoundsError(Exception)>>signal:with:}
	receiver: a BoundsError
	arg[0]: nil
	arg[1]: 47

{1ECE0AD4: cf 1ECE0AB5, sp 1ECE0AE8, bp 1ECE0AD0, ip 5, BoundsError(Exception)>>signalWith:}
	receiver: a BoundsError
	arg[0]: 47

{1ECE0AB4: cf 1ECE0A95, sp 1ECE0AC8, bp 1ECE0AB0, ip 8, ReadStream(Object)>>errorSubscriptBounds:}
	receiver: a ReadStream
	arg[0]: 47

{1ECE0A94: cf 1ECE0A75, sp 1ECE0AA8, bp 1ECE0A90, ip 15, ReadStream(PositionableStream)>>position:}
	receiver: a ReadStream
	arg[0]: 47

{1ECE0A74: cf 1ECE0A45, sp 1ECE0A88, bp 1ECE0A60, ip 13, JadeiteMethodListPresenter>>selecting:string:}
	receiver: a JadeiteMethodListPresenter
	arg[0]: 47
	arg[1]: nil
	stack temp[0]: nil
	stack temp[1]: a ReadStream
	stack temp[2]: nil

{1ECE0A44: cf 1ECE0A21, sp 1ECE0A58, bp 1ECE0A3C, ip 19, JadeiteMethodListPresenter>>updateSingleMethodSource}
	receiver: a JadeiteMethodListPresenter
	stack temp[0]: a RowanMethodService
	stack temp[1]: 47

{1ECE0A20: cf 1ECE0A05, sp 1ECE0A34, bp 1ECE0A20, ip 23, JadeiteMethodListPresenter>>updateMethodSource}
	receiver: a JadeiteMethodListPresenter

{1ECE0A04: cf 1ECE09E1, sp 1ECE0A18, bp 1ECE09FC, ip 13, EventMessageSend>>forwardTo:withArguments:}
	receiver: a EventMessageSend
	arg[0]: a JadeiteMethodListPresenter
	arg[1]: a Array

{1ECE09E0: cf 1ECE09C5, sp 1ECE09F4, bp 1ECE09E0, ip 5, EventMessageSend>>value}
	receiver: a EventMessageSend

{1ECE09C4: cf 1ECE099D, sp 1ECE09D8, bp 1ECE09C0, ip 11, [] in MessageSequenceAbstract>>value}
	receiver: nil
	arg[0]: a EventMessageSend

{1ECE099C: cf 1ECE096D, sp 1ECE09B8, bp 1ECE0988, ip 30, EventMessageSequence>>messagesDo:}
	receiver: a EventMessageSequence
	arg[0]: [] @ 9 in MessageSequenceAbstract>>value
	stack temp[0]: a DeadObject
	stack temp[1]: 2
	stack temp[2]: a JadeiteMethodListPresenter
	stack temp[3]: a EventMessageSend

{1ECE096C: cf 1ECE0951, sp 1ECE0980, bp 1ECE096C, ip 15, EventMessageSequence(MessageSequenceAbstract)>>value}
	receiver: a EventMessageSequence
	env temp[0]: a JadeiteMethodListPresenter

{1ECE0950: cf 1ECE0931, sp 1ECE0964, bp 1ECE094C, ip 8, EventsCollection>>triggerEvent:}
	receiver: a EventsCollection
	arg[0]: methodSelectionChanged

{1ECE0930: cf 1ECE0911, sp 1ECE0944, bp 1ECE092C, ip 4, JadeiteBasicMethodListPresenter(Object)>>trigger:}
	receiver: a JadeiteBasicMethodListPresenter
	arg[0]: methodSelectionChanged

{1ECE0910: cf 1ECE08F5, sp 1ECE0924, bp 1ECE0910, ip 4, JadeiteBasicMethodListPresenter>>methodSelectionChanged}
	receiver: a JadeiteBasicMethodListPresenter

{1ECE08F4: cf 1ECE08D1, sp 1ECE0908, bp 1ECE08EC, ip 13, EventMessageSend>>forwardTo:withArguments:}
	receiver: a EventMessageSend
	arg[0]: a JadeiteBasicMethodListPresenter
	arg[1]: a Array

{1ECE08D0: cf 1ECE08B5, sp 1ECE08E4, bp 1ECE08D0, ip 5, EventMessageSend>>value}
	receiver: a EventMessageSend

{1ECE08B4: cf 1ECE088D, sp 1ECE08C8, bp 1ECE08B0, ip 11, [] in MessageSequenceAbstract>>value}
	receiver: nil
	arg[0]: a EventMessageSend

{1ECE088C: cf 1ECE085D, sp 1ECE08A8, bp 1ECE0878, ip 30, EventMessageSequence>>messagesDo:}
	receiver: a EventMessageSequence
	arg[0]: [] @ 9 in MessageSequenceAbstract>>value
	stack temp[0]: a DeadObject
	stack temp[1]: 1
	stack temp[2]: a JadeiteBasicMethodListPresenter
	stack temp[3]: a EventMessageSend

{1ECE085C: cf 1ECE0841, sp 1ECE0870, bp 1ECE085C, ip 15, EventMessageSequence(MessageSequenceAbstract)>>value}
	receiver: a EventMessageSequence
	env temp[0]: nil

{1ECE0840: cf 1ECE0821, sp 1ECE0854, bp 1ECE083C, ip 8, EventsCollection>>triggerEvent:}
	receiver: a EventsCollection
	arg[0]: selectionChanged

{1ECE0820: cf 1ECE0801, sp 1ECE0834, bp 1ECE081C, ip 4, ListPresenter(Object)>>trigger:}
	receiver: a ListPresenter
	arg[0]: selectionChanged

{1ECE0800: cf 1ECE07E5, sp 1ECE0814, bp 1ECE0800, ip 4, ListView(ListControlView)>>onSelectionChanged}
	receiver: a ListView

{1ECE07E4: cf 1ECE07C9, sp 1ECE07F8, bp 1ECE07E4, ip 3, ListPresenter(SelectableItemsPresenter)>>onSelectionChanged}
	receiver: a ListPresenter

{1ECE07C8: cf 1ECE07A9, sp 1ECE07DC, bp 1ECE07C4, ip 6, ListView(ListControlView)>>onSelChanged:}
	receiver: a ListView
	arg[0]: a Array

{1ECE07A8: cf 1ECE0789, sp 1ECE07BC, bp 1ECE07A4, ip 7, ListView>>onSelChanged:}
	receiver: a ListView
	arg[0]: a Array

{1ECE0788: cf 1ECE0769, sp 1ECE079C, bp 1ECE0784, ip 9, ListView(ListControlView)>>onSelChanged}
	receiver: a ListView
	stack temp[0]: a Array

{1ECE0768: cf 1ECE0745, sp 1ECE077C, bp 1ECE0760, ip 22, ListView>>maybeSelChanging:}
	receiver: a ListView
	arg[0]: mouse
	stack temp[0]: a Array

{1ECE0744: cf 1ECE0725, sp 1ECE0758, bp 1ECE0740, ip 6, ListView>>nmClick:}
	receiver: a ListView
	arg[0]: a ExternalAddress

{1ECE0724: cf 1ECE0701, sp 1ECE0738, bp 1ECE071C, ip 14, ListView(ControlView)>>nmNotify:}
	receiver: a ListView
	arg[0]: a ExternalAddress
	stack temp[0]: nmClick:

{1ECE0700: cf 1ECE06DD, sp 1ECE0714, bp 1ECE06F8, ip 31, ListView>>nmNotify:}
	receiver: a ListView
	arg[0]: a ExternalAddress
	stack temp[0]: -2

{1ECE06DC: cf 1ECE06AD, sp 1ECE06F0, bp 1ECE06C8, ip 18, ContainerView(View)>>wmNotify:wParam:lParam:}
	receiver: a ContainerView
	arg[0]: 78
	arg[1]: 5021
	arg[2]: 1701452
	stack temp[0]: a ExternalAddress
	stack temp[1]: a ListView

{1ECE06AC: cf 1ECE067D, sp 1ECE06C0, bp 1ECE0698, ip 23, ContainerView(View)>>dispatchMessage:wParam:lParam:}
	receiver: a ContainerView
	arg[0]: 78
	arg[1]: 5021
	arg[2]: 1701452
	stack temp[0]: nil
	stack temp[1]: wmNotify:wParam:lParam:

{1ECE067C: cf 1ECE0649, sp 1ECE0690, bp 1ECE066C, ip 69, [] in InputState>>wndProc:message:wParam:lParam:cookie:}
	receiver: nil

{1ECE0648: cf 1ECE0621, sp 1ECE065C, bp 1ECE0644, ip 18, BlockClosure>>ifCurtailed:}
	receiver: [] @ 0 in nil
	arg[0]: [] @ 79 in InputState>>wndProc:message:wParam:lParam:cookie:

{1ECE0620: cf 1ECE05D5, sp 1ECE063C, bp 1ECE0608, ip 84, InputState>>wndProc:message:wParam:lParam:cookie:}
	receiver: a InputState
	arg[0]: 5309948
	arg[1]: 78
	arg[2]: 5021
	arg[3]: 1701452
	arg[4]: 850310
	stack temp[0]: a ContainerView

{1ECE05D4: cf 1ECE05AD, sp 1ECE0600, bp 1ECE05C8, ip 10, ListView(ControlView)>>defaultWindowProcessing:wParam:lParam:}
	receiver: a ListView
	arg[0]: 513
	arg[1]: 1
	arg[2]: 2294109

{1ECE05AC: cf 1ECE0591, sp 1ECE05C0, bp 1ECE05AC, ip 12, MouseEvent(WindowsEvent)>>defaultWindowProcessing}
	receiver: a MouseEvent

{1ECE0590: cf 1ECE0569, sp 1ECE05A4, bp 1ECE0584, ip 17, ListView>>onButtonPressed:}
	receiver: a ListView
	arg[0]: a MouseEvent
	stack temp[0]: a Array
	stack temp[1]: nil

{1ECE0568: cf 1ECE0549, sp 1ECE057C, bp 1ECE0564, ip 8, ListView>>onLeftButtonPressed:}
	receiver: a ListView
	arg[0]: a MouseEvent

{1ECE0548: cf 1ECE0529, sp 1ECE055C, bp 1ECE0544, ip 4, ListPresenter(Presenter)>>onLeftButtonPressed:}
	receiver: a ListPresenter
	arg[0]: a MouseEvent

{1ECE0528: cf 1ECE0501, sp 1ECE053C, bp 1ECE051C, ip 9, ListView(View)>>wmLButtonDown:wParam:lParam:}
	receiver: a ListView
	arg[0]: 513
	arg[1]: 1
	arg[2]: 2294109

{1ECE0500: cf 1ECE04D1, sp 1ECE0514, bp 1ECE04EC, ip 23, ListView(View)>>dispatchMessage:wParam:lParam:}
	receiver: a ListView
	arg[0]: 513
	arg[1]: 1
	arg[2]: 2294109
	stack temp[0]: nil
	stack temp[1]: wmLButtonDown:wParam:lParam:

{1ECE04D0: cf 1ECE049D, sp 1ECE04E4, bp 1ECE04C0, ip 69, [] in InputState>>wndProc:message:wParam:lParam:cookie:}
	receiver: nil

{1ECE049C: cf 1ECE0475, sp 1ECE04B0, bp 1ECE0498, ip 18, BlockClosure>>ifCurtailed:}
	receiver: [] @ 0 in nil
	arg[0]: [] @ 79 in InputState>>wndProc:message:wParam:lParam:cookie:

{1ECE0474: cf 1ECE0435, sp 1ECE0490, bp 1ECE045C, ip 84, InputState>>wndProc:message:wParam:lParam:cookie:}
	receiver: a InputState
	arg[0]: 1641266
	arg[1]: 513
	arg[2]: 1
	arg[3]: 2294109
	arg[4]: 851230
	stack temp[0]: a ListView

{1ECE0434: cf 1ECE0415, sp 1ECE0454, bp 1ECE0430, ip 6, ShellView(View)>>isDialogMessage:}
	receiver: a ShellView
	arg[0]: a MSG

{1ECE0414: cf 1ECE03F5, sp 1ECE0428, bp 1ECE0410, ip 5, ShellView>>preTranslateMouseInput:}
	receiver: a ShellView
	arg[0]: a MSG

{1ECE03F4: cf 1ECE03D5, sp 1ECE0408, bp 1ECE03F0, ip 13, ShellView(View)>>preTranslateMessage:}
	receiver: a ShellView
	arg[0]: a MSG

{1ECE03D4: cf 1ECE03AD, sp 1ECE03E8, bp 1ECE03C8, ip 17, InputState>>preTranslateMessage:}
	receiver: a InputState
	arg[0]: a MSG
	stack temp[0]: 9241674
	stack temp[1]: a ShellView

{1ECE03AC: cf 1ECE038D, sp 1ECE03C0, bp 1ECE03A8, ip 19, InputState>>pumpMessage:}
	receiver: a InputState
	arg[0]: a MSG

{1ECE038C: cf 1ECE0361, sp 1ECE03A0, bp 1ECE037C, ip 18, InputState>>loopWhile:}
	receiver: a InputState
	arg[0]: [] @ 9 in InputState>>mainLoop
	stack temp[0]: a MSG
	stack temp[1]: true
	stack temp[2]: nil

{1ECE0360: cf 1ECE0345, sp 1ECE0374, bp 1ECE0360, ip 15, InputState>>mainLoop}
	receiver: a InputState

{1ECE0344: cf 1ECE0329, sp 1ECE0358, bp 1ECE0344, ip 14, [] in InputState>>forkMain}
	receiver: a InputState

{1ECE0328: cf 1ECE030D, sp 1ECE033C, bp 1ECE0328, ip 13, ExceptionHandler(ExceptionHandlerAbstract)>>markAndTry}
	receiver: a ExceptionHandler

{1ECE030C: cf 1ECE02E5, sp 1ECE0320, bp 1ECE0308, ip 22, [] in ExceptionHandler(ExceptionHandlerAbstract)>>try:}
	receiver: a ExceptionHandler

{1ECE02E4: cf 1ECE02C5, sp 1ECE02F8, bp 1ECE02E0, ip 18, BlockClosure>>ifCurtailed:}
	receiver: [] @ 0 in nil
	arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:

{1ECE02C4: cf 1ECE02A1, sp 1ECE02D8, bp 1ECE02BC, ip 3, BlockClosure>>ensure:}
	receiver: [] @ 17 in ExceptionHandlerAbstract>>try:
	arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:
	stack temp[0]: nil

{1ECE02A0: cf 1ECE0279, sp 1ECE02B4, bp 1ECE0294, ip 44, ExceptionHandler(ExceptionHandlerAbstract)>>try:}
	receiver: a ExceptionHandler
	arg[0]: [] @ 10 in InputState>>forkMain
	stack temp[0]: nil
	stack temp[1]: a Process('Main' base 1ECE0000 [ACTIVE] in SessionManager>>logError: sp=00000000 ip=4 list=02770010)
	env temp[0]: nil

{1ECE0278: cf 1ECE0255, sp 1ECE028C, bp 1ECE0270, ip 7, BlockClosure>>on:do:}
	receiver: [] @ 10 in InputState>>forkMain
	arg[0]: ProcessTermination
	arg[1]: [] @ 13 in BlockClosure>>newProcess

{1ECE0254: cf 00000001, sp 1ECE0268, bp 1ECE0254, ip 17, [] in BlockClosure>>newProcess}
	receiver: [] @ 10 in InputState>>forkMain

<Bottom of stack>

***** End of dump *****

***** Ending Dolphin stack dump started at 11:07:23 AM 3/4/2020 *****
```

**Comment by ericwinger** [comment]
Do you remember how you opened this particular method list browser, @dalehenrich ?

**Comment by dalehenrich** [comment]
I think it was with either senders or implementers ... most likely senders...

**Comment by ericwinger** [comment]
@dalehenrich The code is trying to set the selection range in the method source pane of the first occurrence of the sender/implementer. It thinks the position of the start of the string is at 47, but (from observation of your window), finds an empty method source pane. The source pane should never be empty therefore it's possible there was an earlier error. 

Do you still have the full `jadeite.errors` file around?  If so, can you attach it to this issue? 
(In lieu of copy/pasting the whole file ... If you rename the file to `jadeite.errors.txt` you can drag it on to the github comment field which will create a link to the file.)

**Comment by ericwinger** [comment]
@dalehenrich  hold off for a second.  I think I just found a reproduction case.

**Comment by ericwinger** [comment]
The reproduction case was not reliable but I was able to get one live Dolphin debugger on the problem. 

When a method is selected in the method list browser, a number of events are triggered. One event calls the server to update the method. Depending on the order of these events it is possible, although not probable, that the source pane will not be finished updating before Jadeite tries to highlight the text range of the first sender. 

You can get a hint of the problem by double-clicking quickly on a new method in the method list browser. A double click opens the project browser on the selected method. Most of the time, the source pane changes to the new method source before opening the project browser but, on occasion, one sees that the source pane becomes blank while the project browser is opening. The method source pane is updated after the project browser is open. 

The best workaround is to skip the text highlighting if the source pane has not been fully updated. It is better than an error. 

Fixed in Oscar 3.1.0
